[onechat] Update tools UX for tools landing page

dennis-tismenko opened the pull request:
## Summary

Overhauls the UX for the tools landing page to align with the latest UX designs ([figma](https://www.figma.com/design/GAFnrYGIEbk7nzIOy9PSso/OneChat-Milestone-2---Tools?node-id=157-13880&p=f&m=dev)), including:
- (Re-)Unifying system and custom tools tables
- Local storage persistence for including/excluding system tools
- Filtering on tool type and labels
- Bulk actions for select and delete, including new `POST _bulk_delete` endpoint
- Row quick actions for custom tools
- Row context menu for all tools
- Implement "Clone tool" 

**Out of scope**
- Changes to create and edit views, including moving from a flyout to separate page

#### Demo

https://github.com/user-attachments/assets/c836bb28-80ba-4576-a5ae-2760b139b3bb


### Checklist

Check the PR satisfies following conditions. 

Reviewers should verify this PR satisfies this list as well.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/src/platform/packages/shared/kbn-i18n/README.md)
- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [x] This was checked for breaking HTTP API changes, and any breaking changes have been approved by the breaking-change committee. The `release_note:breaking` label should be applied in these situations.
- [x] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [x] The PR  description includes the appropriate Release Notes section, and the correct `release_note:*` label is applied per the [guidelines](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)
- [x] Review the [backport guidelines](https://docs.google.com/document/d/1VyN5k91e5OVumlc0Gb9RPa3h1ewuPE705nRtioPiTvY/edit?usp=sharing) and apply applicable `backport:*` labels.

dennis-tismenko committed:
Refactor tools page for v2 UX - WIP

- Update header copy
- Remove header background
- Remove border between header and page body
- Move "New tool" button to header
- Update copy in "New tool" button
- Make search bar full width
- Unify tools table

dennis-tismenko committed:
Refactor tools page for v2 UX - WIP

- Pagination text
- Update column names and content

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

dennis-tismenko committed:
Refactor tools page for v2 UX - WIP

- Context menu
- Quick actions on hover
- Bulk actions (selection only)

dennis-tismenko committed:
Refactor tools page for v2 UX - WIP

- Bulk actions: Select all and clear
- Bulk actions: Delete selected (UI ONLY)
- Type filter
- Label filter

dennis-tismenko committed:
Refactor tools page for v2 UX - WIP

- Bulk delete

dennis-tismenko committed:
impl: Add "Include system tools" toggle

dennis-tismenko committed:
impl: Make system tools toggle persistent in localStorage

dennis-tismenko committed:
fix: tags not filtering out system tool tags

dennis-tismenko committed:
impl: custom tool cloning

dennis-tismenko committed:
impl: add filter counts

dennis-tismenko committed:
refactor: remove underline on hover

dennis-tismenko committed:
refactor: clean up implementation with separation of components and hooks

dennis-tismenko committed:
refactor: use context for tools actions

dennis-tismenko committed:
impl: Better bulk delete API, FTR tests, bug fixes in UI

dennis-tismenko committed:
refactor: self code review

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/327122)
* Commit: 6979db0a21f88d0d30fd7456a716b4b90ca61a01

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/327122#0198800c-9146-46a4-89fb-d77620bb5f50)

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/327122","number":327122,"commit":"6979db0a21f88d0d30fd7456a716b4b90ca61a01"}],"number":327122}
buildkite-pr-comment-->

dennis-tismenko committed:
fix: duplicate i18n ids

dennis-tismenko committed:
fix: remove useTools dependency on ToolsPreferencesProvider

dennis-tismenko committed:
fix: failing FTR test

jedrazb commented on x-pack/platform/plugins/shared/onechat/public/application/components/tools/table/tools_table_search.tsx:
```diff
@@ -0,0 +1,138 @@
+  }, [tools]);
+
+  const matchesByTag = useMemo(() => {
+    return countBy(tools, 'tags');
```
sth wrong here, tags is an array will countby still work for multivalued fields?

see screenshot I feel like it should be 3 and 1 but it's showing 2 and 0

<img width="1392" height="580" alt="Image" src="https://github.com/user-attachments/assets/aed5f790-947f-44bf-a172-0306957cf107" />

dennis-tismenko replied:
Right, for some reason I thought it'd flatten the list but it just stores the collection as a key. Will fix by doing a `flatMap` on the tags before counting:

```suggestion
    return countBy(tools.flatMap((tool) => tool.tags));
```

jedrazb commented on x-pack/platform/plugins/shared/onechat/public/application/context/tools_provider.tsx:
```diff
@@ -0,0 +1,165 @@
+
+export const ToolsActionsContext = createContext<ToolsActionsContextType | undefined>(undefined);
+
+export const ToolsProvider = ({ children }: { children: React.ReactNode }) => {
```
that's a beefy provider, we probably will be able to split this up once we move to per-tool page create/edit page?

dennis-tismenko replied:
100%, but as beefy as it is, I think it still make sense so that any component within the provider context can trigger an action that opens a modal/flyout as needed, but I'm also open to other ways of splitting this up to achieve a similar effect.

Once we move to pages from flyouts, I'll get rid of all of the flyout rendering logic and `createTool` and `editTool` actions will be left here but converted to redirects with `useNavigation`. I'm thinking of keeping modal rendering + delete hooks here though, thoughts on this?

Might also be worth renaming to something like `ToolsTableProvider` so that its scope is more clear, just thinking out loud.

jedrazb commented on x-pack/platform/plugins/shared/onechat/server/routes/tools.ts:
```diff
@@ -227,7 +229,62 @@ export function registerToolsRoutes({ router, getInternalServices, logger }: Rou
+  // bulk delete tools
+  router.versioned
+    .post({
+      path: '/api/chat/tools/_bulk_delete',
```
do we really want to expose this tools-specific _bulk_delete endpoint as public, it's not part of the onechat api spec

maybe keep this as internal endpoint, @pgayvallet @joemcelroy wdyt>

jedrazb commented on x-pack/platform/plugins/shared/onechat/public/application/components/tools/table/tools_table_search.tsx:
```diff
@@ -0,0 +1,138 @@
+import { ToolFilterOption } from './tools_table_filter_option';
+import { useOnechatTools } from '../../../hooks/tools/use_tools';
+
+const getToolsTableSearchConfig = ({
```
should we consider keeping search state in URL? E.g. to allow navigation to show only tools with given tag etc. 

I don't like in-mem search states in webpages as you can't really share / save its result

dennis-tismenko replied:
Yeah, that's a great idea. I'll implement this

dennis-tismenko replied:
Actually, I'm not sure it makes sense for this use case — is being able to save and share tool table states relevant for the tools page? I 100% agree it usually makes sense for more complex data tables, just not fully convinced it's needed here.

If we still want to go down this route, would we want searches to update the browser history stack (ie. pressing back will take you to a previous search, as opposed to the previous page we're on)?

jedrazb reviewed (commented):
a bug with tags counting, couple of questions! will do a 2nd pass tomorrow

dennis-tismenko committed:
fix: review suggestions

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/notice'

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/327815)
* Commit: 06d0abddad1d256803348e9aa95f0fad2abfe6d1

### Failed CI Steps
* [Jest Tests #2](https://buildkite.com/elastic/kibana-pull-request/builds/327815#019888e8-8d38-4cbf-a1ec-125483d12bff)
* [Jest Tests #2](https://buildkite.com/elastic/kibana-pull-request/builds/327815#0198890f-6649-40aa-87fe-63670b63293c)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/327815#019888e8-8d38-4cbf-a1ec-125483d12bff) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/327815/jobs/019888e8-8d38-4cbf-a1ec-125483d12bff/artifacts/0198890f-24b5-4649-891f-9dfa5830ac32) Jest Tests #<span></span>2 / bulkFillGapsByRuleIds should call batchBackfillRuleGaps correctly
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/327815#0198890f-6649-40aa-87fe-63670b63293c) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/327815/jobs/0198890f-6649-40aa-87fe-63670b63293c/artifacts/01988922-2b6c-4ca3-9329-705123723ab5) Jest Tests #<span></span>2 / bulkFillGapsByRuleIds should call batchBackfillRuleGaps correctly
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/327815#019888e8-8d38-4cbf-a1ec-125483d12bff) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/327815/jobs/019888e8-8d38-4cbf-a1ec-125483d12bff/artifacts/0198890f-24b2-41ae-acc9-f1a03df98ecd) Jest Tests #<span></span>2 / bulkFillGapsByRuleIds should ensure the user is authorized for each combination of ruleTypeId and consumer found in the rules
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/327815#0198890f-6649-40aa-87fe-63670b63293c) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/327815/jobs/0198890f-6649-40aa-87fe-63670b63293c/artifacts/01988922-2b6d-495c-a02d-a475d88725cd) Jest Tests #<span></span>2 / bulkFillGapsByRuleIds should ensure the user is authorized for each combination of ruleTypeId and consumer found in the rules
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/327815#019888e8-8d38-4cbf-a1ec-125483d12bff) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/327815/jobs/019888e8-8d38-4cbf-a1ec-125483d12bff/artifacts/0198890f-24b6-42bf-b754-0da29319ee34) Jest Tests #<span></span>2 / bulkFillGapsByRuleIds should log an audit event when the rule fails at the authorization step
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/327815#0198890f-6649-40aa-87fe-63670b63293c) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/327815/jobs/0198890f-6649-40aa-87fe-63670b63293c/artifacts/01988922-2b6c-4be4-a522-eda5167ca256) Jest Tests #<span></span>2 / bulkFillGapsByRuleIds should log an audit event when the rule fails at the authorization step
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/327815#019888e8-8d38-4cbf-a1ec-125483d12bff) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/327815/jobs/019888e8-8d38-4cbf-a1ec-125483d12bff/artifacts/0198890f-24b4-4b20-9c79-be5d6013ce0c) Jest Tests #<span></span>2 / bulkFillGapsByRuleIds should refresh the index of the event client once after scheduling the backfilling of all gaps
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/327815#0198890f-6649-40aa-87fe-63670b63293c) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/327815/jobs/0198890f-6649-40aa-87fe-63670b63293c/artifacts/01988922-2b6c-4b57-8800-5ce1e6c5bd6f) Jest Tests #<span></span>2 / bulkFillGapsByRuleIds should refresh the index of the event client once after scheduling the backfilling of all gaps
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/327815#019888e8-8d38-4cbf-a1ec-125483d12bff) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/327815/jobs/019888e8-8d38-4cbf-a1ec-125483d12bff/artifacts/0198890f-24b3-4212-93f7-31775f7ca9b6) Jest Tests #<span></span>2 / bulkFillGapsByRuleIds should return a list with errored rules
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/327815#0198890f-6649-40aa-87fe-63670b63293c) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/327815/jobs/0198890f-6649-40aa-87fe-63670b63293c/artifacts/01988922-2b6d-4707-9a16-85c7f3d2c671) Jest Tests #<span></span>2 / bulkFillGapsByRuleIds should return a list with errored rules
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/327815#019888e8-8d38-4cbf-a1ec-125483d12bff) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/327815/jobs/019888e8-8d38-4cbf-a1ec-125483d12bff/artifacts/0198890f-24b5-442f-8848-d5d6b717894b) Jest Tests #<span></span>2 / bulkFillGapsByRuleIds should return a list with rules that succeeded
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/327815#0198890f-6649-40aa-87fe-63670b63293c) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/327815/jobs/0198890f-6649-40aa-87fe-63670b63293c/artifacts/01988922-2b6d-407b-be96-22cf163337b4) Jest Tests #<span></span>2 / bulkFillGapsByRuleIds should return a list with rules that succeeded
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/327815#019888e8-8d38-4cbf-a1ec-125483d12bff) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/327815/jobs/019888e8-8d38-4cbf-a1ec-125483d12bff/artifacts/0198890f-24b4-4818-b3d8-a7aaf48649d1) Jest Tests #<span></span>2 / bulkFillGapsByRuleIds should return a list with rules that were skipped
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/327815#0198890f-6649-40aa-87fe-63670b63293c) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/327815/jobs/0198890f-6649-40aa-87fe-63670b63293c/artifacts/01988922-2b6d-4ad8-bdec-28fd9cd9a993) Jest Tests #<span></span>2 / bulkFillGapsByRuleIds should return a list with rules that were skipped

### History
* :yellow_heart: [Build #327222](https://buildkite.com/elastic/kibana-pull-request/builds/327222) was flaky 75db9353aebed1bbc5b8591f58b3b06f9aa419ba
* :broken_heart: [Build #327157](https://buildkite.com/elastic/kibana-pull-request/builds/327157) failed ff86d18e034d9a294f3235005555084b6c853cde
* :broken_heart: [Build #327122](https://buildkite.com/elastic/kibana-pull-request/builds/327122) failed 6979db0a21f88d0d30fd7456a716b4b90ca61a01

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/327815","number":327815,"commit":"06d0abddad1d256803348e9aa95f0fad2abfe6d1"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/327222","number":327222,"commit":"75db9353aebed1bbc5b8591f58b3b06f9aa419ba"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/327157","number":327157,"commit":"ff86d18e034d9a294f3235005555084b6c853cde"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/327122","number":327122,"commit":"6979db0a21f88d0d30fd7456a716b4b90ca61a01"}],"number":327815}
buildkite-pr-comment-->

dennis-tismenko committed:
refactor: ToolsProvider to ToolsTableProvider

dennis-tismenko committed:
fix: errors when parsing malformed search queries in the URL

dennis-tismenko commented:
> a bug with tags counting, couple of questions! will do a 2nd pass tomorrow

Fixed, and addressed the other feedback

dennis-tismenko committed:
refactor: remove effect that isn't really doing anything

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/328046)
* Commit: d267f780102862ff30b01776fcc72a273553a439

### Failed CI Steps
* [FTR Configs #132](https://buildkite.com/elastic/kibana-pull-request/builds/328046#01988aad-c513-426b-8bdd-783ba9b1c0b5)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/328046#01988aad-c513-426b-8bdd-783ba9b1c0b5) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/328046/jobs/01988aad-c513-426b-8bdd-783ba9b1c0b5/artifacts/01988ad2-3df4-4270-8ec5-b200450d186c) FTR Configs #<span></span>132 / InfraOps App feature controls infrastructure security global infrastructure read privileges infrastructure landing page with data shows Wafflemap

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/b70f455d2fcb3640b8be4fb6c7aecff8e2f7e161) | [after](https://github.com/elastic/kibana/commit/d267f780102862ff30b01776fcc72a273553a439) | diff |
| --- | --- | --- | --- |
| `onechat` | 418 | 428 | +10 |

#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/b70f455d2fcb3640b8be4fb6c7aecff8e2f7e161) | [after](https://github.com/elastic/kibana/commit/d267f780102862ff30b01776fcc72a273553a439) | diff |
| --- | --- | --- | --- |
| `@kbn/onechat-common` | 141 | 142 | +1 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/b70f455d2fcb3640b8be4fb6c7aecff8e2f7e161) | [after](https://github.com/elastic/kibana/commit/d267f780102862ff30b01776fcc72a273553a439) | diff |
| --- | --- | --- | --- |
| `onechat` | 769.0KB | 780.2KB | +11.2KB |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/b70f455d2fcb3640b8be4fb6c7aecff8e2f7e161) | [after](https://github.com/elastic/kibana/commit/d267f780102862ff30b01776fcc72a273553a439) | diff |
| --- | --- | --- | --- |
| `onechat` | 12.2KB | 12.6KB | +329.0B |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/b70f455d2fcb3640b8be4fb6c7aecff8e2f7e161) | [after](https://github.com/elastic/kibana/commit/d267f780102862ff30b01776fcc72a273553a439) | diff |
| --- | --- | --- | --- |
| `@kbn/onechat-common` | 218 | 219 | +1 |


</details>


### History
* :broken_heart: [Build #327815](https://buildkite.com/elastic/kibana-pull-request/builds/327815) failed 06d0abddad1d256803348e9aa95f0fad2abfe6d1
* :yellow_heart: [Build #327222](https://buildkite.com/elastic/kibana-pull-request/builds/327222) was flaky 75db9353aebed1bbc5b8591f58b3b06f9aa419ba
* :broken_heart: [Build #327157](https://buildkite.com/elastic/kibana-pull-request/builds/327157) failed ff86d18e034d9a294f3235005555084b6c853cde
* :broken_heart: [Build #327122](https://buildkite.com/elastic/kibana-pull-request/builds/327122) failed 6979db0a21f88d0d30fd7456a716b4b90ca61a01

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/328046","number":328046,"commit":"d267f780102862ff30b01776fcc72a273553a439"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/327815","number":327815,"commit":"06d0abddad1d256803348e9aa95f0fad2abfe6d1"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/327222","number":327222,"commit":"75db9353aebed1bbc5b8591f58b3b06f9aa419ba"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/327157","number":327157,"commit":"ff86d18e034d9a294f3235005555084b6c853cde"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/327122","number":327122,"commit":"6979db0a21f88d0d30fd7456a716b4b90ca61a01"}],"number":328046}
buildkite-pr-comment-->

jedrazb reviewed (approved):
LGTM, I hope much of the tools table provider will get simplified once we break this down into separate pages. Please double check the internal bulk delete api once Joe and Pierre are back
